### ISSUE TITLE
gettext build broken on substituted strings

### ISSUE THREAD

**Issue by doctorlard** [sphinx-doc/sphinx#3093]
When running:

```
  sphinx-build -b gettext ...
```

In `sphinx/builders/gettext.py` MessageCatalogBuilder.finish() calls canon_path() which throws an AttributeError when its argument is None, which seems to happen in situations with lots of `.. image::` directives with `:alt:` messages. This has been broken since at least October last year (commit 281bf2a).
I could fix the functions, but this will probably just hide the possible bug somewhere else that is returning None for the translation string source file and line number.

**Comment by doctorlard** [comment]
It appears that using substitution strings causes the gettext builder to lose track of the source file and line number, which results in canon_path being sent None instead of a string.

**Comment by tk0miya** [comment]
Could you show me some reproducible example?

**Comment by doctorlard** [comment]
Our manual code is here: https://git.mahara.org/user-manual/manual/tree/16.10_STABLE
Results are generated here: http://manual.mahara.org/en/16.10/

**Comment by tk0miya** [comment]
Thanks! I can reproduce the problem. I will check it!

```
# Sphinx version: 1.4.8
# Python version: 2.7.12 (CPython)
# Docutils version: 0.12 release
# Jinja2 version: 2.8
# Last messages:
#   writing output... [ 89%] portfolio/skins
#   writing output... [ 91%] portfolio/smartevidence
#   writing output... [ 92%] portfolio/tags
#   writing output... [ 94%] staff/staff
#   writing output... [ 95%] tips/image_in_text
#   writing output... [ 97%] tips/mahoodle_deep_link
#   writing output... [ 98%] tips/tips
#   writing output... [100%] todo
#
#   writing message catalogs... [  6%] index
# Loaded extensions:
#   sphinxjp.themes.basicstrap (unknown version) from /Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinxjp/themes/basicstrap/__init__.pyc
#   sphinx.ext.todo (1.4.8) from /Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/ext/todo.pyc
#   sphinx.ext.autodoc (1.4.8) from /Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/ext/autodoc.pyc
#   sphinx.ext.ifconfig (1.4.8) from /Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/ext/ifconfig.pyc
#   alabaster (0.7.9) from /Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/alabaster/__init__.pyc
Traceback (most recent call last):
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/cmdline.py", line 244, in main
    app.build(opts.force_all, filenames)
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/application.py", line 297, in build
    self.builder.build_update()
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 251, in build_update
    'out of date' % len(to_build))
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 195, in build
    I18nBuilder.build(self, docnames, summary, method)
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 325, in build
    self.finish()
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 226, in finish
    for source, line, _ in positions))
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 226, in <genexpr>
    for source, line, _ in positions))
  File "/Users/tkomiya/work/tmp/manual-16.10_STABLE-e31bb451f960ee33a3402796a96c9b1cc417091d/lib/python2.7/site-packages/sphinx/util/osutil.py", line 45, in canon_path
    return nativepath.replace(path.sep, SEP)
AttributeError: 'NoneType' object has no attribute 'replace'
```

**Comment by tk0miya** [comment]
@shimizukawa It seems both source and line are None value. Is this correct?

**Comment by shimizukawa** [comment]
To reproduce the issue:
- conf.py: `gettext_additional_targets = ['image']`
- some.rst: `.. |picture file| image:: foo.*`

It's a bug of Sphinx.

**Comment by shimizukawa** [comment]
I found another bug:
- conf.py: `gettext_additional_targets = ['image']`
- some.rst: 

```
.. note::

   .. image:: foo.png
      :alt: Foo!
```

**Comment by tk0miya** [comment]
Fixed at #3165 
Thank you for reporting!
